### ISSUE TITLE
WebP and TIFF encoders produce invalid images when writing to HttpContext.Response.Body

### ISSUE THREAD

**Issue by csboling** [SixLabors/ImageSharp#2806]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have verified that I am running the latest version of ImageSharp
- [X] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [X] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### ImageSharp version

3.1.5

### Other ImageSharp packages and versions

N/A

### Environment (Operating system, version and so on)

Windows 10 19044.2604, 64-bit

### .NET Framework version

.NET 8

### Description

Encoding an `Image` as WebP to a `Microsoft.AspNetCore.Http.HttpContext` (a non-seekable `Stream`) produces an image that cannot be displayed in the browser (Firefox 129.0.2, Chrome 128.0.6613.114) and has a different length and contents from the same `Image` written to a file. Writing an `Image` as TIFF in the same way produces a file which cannot be displayed in the Windows image viewer and contains no valid IFDs. All other formats currently supported by ImageSharp produce valid images when written in this way. Tested with both `Image<Rgba32>` and `Image<Rgb24>`. I would guess it affects other non-seekable `Stream`s with these formats as well.

### Steps to Reproduce

Here is a code sample reproducing the problem. A complete project can be found here: https://github.com/csboling/imagesharp-webp-http

```c#
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.MapGet(
    "/download.webp",
    async ctx =>
    {
        using (Image<Rgba32> image = new(1, 1))
        {
            ctx.Response.ContentType = "image/webp";
            await image.SaveAsWebpAsync(ctx.Response.Body);
        }
    });

app.Run();
```

### Images

This ZIP contains example 1x1 pixel RGBA32 images encoded to WebP and TIFF to an `HttpContext.Response`, then downloaded with Chrome (invalid.webp, invalid.tif) as well as images encoded directly to a file (valid.webp, valid.tif).

[examples.zip](https://github.com/user-attachments/files/16911652/examples.zip)

**Comment by JimBobSquarePants** [comment]
We explicitly wrap non seekable streams https://github.com/SixLabors/ImageSharp/blob/release/3.1.x/src/ImageSharp/Formats/ImageEncoder.cs

**Comment by SamsidParty** [comment]
I can reproduce this issue when writing to a Zip Output Stream, examining the webp file in a hex editor shows that the header doesn't contain a length value. Using version 2.1.9, it works fine.

**Comment by JimBobSquarePants** [comment]
Ok. Thanks. I’m on holiday in Europe currently don’t have access to a laptop just now so will need someone to debug

**Comment by JimBobSquarePants** [comment]
I've found a few issues with the current `ChunkedMemoryStream` implementation. Rewriting it now.

**Comment by JimBobSquarePants** [comment]
Fixed via #2828
